### ISSUE TITLE
MLQM: KRR not matching with Margraf/Reuter results

### ISSUE THREAD

**Issue by bgpeyton** [bgpeyton/lml#13]
There's still a bug in the KRR code somewhere. `MLQM` reproduced the [Magraf and Reuter](https://pubs.acs.org/doi/10.1021/acs.jpca.8b04455) results for H2 quite well (MP2 or CCSD TATRs), but for other diatomics the regression is not nearly as accurate. I'll try to keep track of things I've tried, changed, or fixed here until I find the problem. 

- Fixed `solve_alpha()`. Now adding the regularization term to the kernel before solving the linear equations, rather than subtract. Error in the Margraf/Reuter Supp Info, confirmed by the scikit-learn(`skl`) code and eq 14.36 in Murphy's "Machine Learning: A Probabilistic Perspective". 
- Tried extending to negative hyperparameter values, no luck. 
- Tried "preferring" high/low hyperparameters when errors are the same. These bounded my regression by over/under training, respectively. Currently preferring higher hypers in the code (see`krr.grid_search()`).
- Changed L2 loss function to use `np.linalg.norm()` instead of calculating the norm manually.
- Tried dropping the regularization from the loss function, no luck.

I've also compared my code deeply with `skl` in order to try troubleshooting the problem. For this I used TATRs generated by my own code and training sets chosen by my k-means algorithm, then pushed them into an `skl.KernelRidge(kernel='rbf')` object.
- When using hyperparameters found by my code, `alpha` (`dual_coef_` in skl) is found to be the same as my code. Same for the kernel. 
- - As the kernel and `alpha` are the same, the predicted results are also the same (which is to say, not matching with Margraf/Reuter).
- When using `skl.GridSearchCV()` over the same grid as `MLQM`, the optimized hyperparameters are different. Tried many different loss functions-- while these did produce different hyperparameters (which, when pushed all the way to the end, gave sometimes slightly better results in the regression), it still did not compare well to the Margraf and Reuter results. None of the loss functions I tried matched `MLQM` either. I have not yet tried adapting some `skl` loss functions into `MLQM`.
- The default loss function in `skl` (R2 loss) shouldn't even be valid for the case of leave-one-out cross-validation, which I assume is the default for `skl.GridSearchCV()` (see documentation). Not sure why that's not throwing an error as it should, as even directly specifying `GridSearchCV(krr,parameters,scoring='r2')` does not throw an error. (NOTE: try specifying k-fold CV, where k = M!)
- - UPDATE: forced k = M, and now the R2 loss function is throwing an error (as expected). This means the default was _not_ LOO-CV. While this improved results (for non-R2 loss functions), it did not recover Margraf/Reuter accuracy.
- The k-means algorithm in `skl` gives slightly different results than that in `MLQM`. The geometry metric they use to determine the best clustering is the "inertia" or, "sum of squared distances to the closest centroid for all observations in the training set." 
- - from their [code](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/cluster/_k_means.pyx#L77): 
- - hardcoded: minimize euclidean distance to cluster center: `||a - b||^2 = ||a||^2 + ||b||^2 -2 <a, b>`
- - tried changing my code to use this metric, but still seem to be getting different clusterings. may try to force equivalent seeds next. 

The above findings lead me to believe there is an error in one of two places: the input data (ie the energies, representations, or ID of training set) or the cross-validation (ie the loss function or grid search). If the former is the case, it is purely a problem in `MLQM` (such as: improper clustering or TATR generation). If the latter, then there is an error in my `skl` notebook (such as: not having the proper loss function, or searching over too small/large a grid in the CV step).

**Comment by bgpeyton** [comment]
UPDATE: forced k = M, and now the R2 loss function is throwing an error (as expected). This means the default was _not_ LOO-CV. While this improved results, it did not recover Margraf/Reuter accuracy.

**Comment by bgpeyton** [comment]
UPDATE: upon getting some extra data for comparison from Margraf (geometries and corresponding energies + amplitudes for the CO PES), it appears that my amplitudes are only slightly off from his once using the exact same PES (~1e-6). Symmetry seems to be conserved the same as my code (as in: highest `abs()` of all amplitudes, not just the unique-by-symmetry ones). 

He also mentioned that he builds his TATRs to be normalized such that they integrate to one. Using this representation, he found that sigma=2048.0 and lambda=1e-06 for the case of 12 training points. Using this information, I _may_ be able to tease out some information on where my code is going wrong.

**Comment by bgpeyton** [comment]
Solution: 

It seems the amplitudes are from calculations utilizing molecular symmetry. In this case, the set of 150 amplitudes is (for some reason) a better description of the wavefunction for machine learning purposes. 

The differences in the hypers should come down to different normalization of the TATRs, and the k-means differences will be evaluated elsewhere.
